0000-catalog.md

```diff
@@ -0,0 +1,801 @@
+    string catalog_id = 1;
+    string product_id = 2;
+}
+```
+
+If the Grid setting `grid.product.allow_delete` is set to false, this transaction
+is invalid (sys admin setting). The default value for `grid.product.allow_delete`
+is true. This setting is stored using the Sawtooth Settings smart contract, more 
+information can be found [here](https://sawtooth.hyperledger.org/docs/core/releases/1.0/transaction_family_specifications/settings_transaction_family.html).
+
+Validation requirements:
+
+- If a catalog_product with catalog_product_id does not exist, the transaction is
+  invalid.
+- The signer of the transaction must be an agent in the Pike state and must
+  belong to an organization in Pike state, otherwise the transaction is invalid.
+- The owner in the product must match the organization that the agent belongs
+  to, otherwise the transaction is invalid.
+- The agent must have the permission “can_delete_product” for the organization
+  otherwise the transaction is invalid.
+
+The inputs for CatalogProductDeleteAction must include:
+
+- Address of the Agent submitting the transaction
+- Address of the Organization the Product is being deleted for
+- Address of the catalog_product to be deleted
+
+The outputs for CatalogProductDeleteAction must include:
+
+- Address of the catalog_product to be deleted
+
+**_NOTE: Deleting a catalog_product is potentially dangerous operation that
+could leave dangling references and should be done with care._**
+
+## Catalog_Product Operations
+
+### CatalogProductSetStatus
+
+CatalogProductSetStatusAction updates an existing catalog_product in state. The
+transaction should be submitted by an agent, identified by its signing key, 
+acting on behalf of an organization that corresponds to the owner in the 
+product being updated. (Organizations and agents are defined by the Pike smart 
+contract.)
+
+The implementation of the CatalogProductSetStatusAction will handle iterating 
+through the list of catalog_ids, and update the status of the 
+catalog_product(s) accordingly. Having the address of an individual 
+catalog_product be a composite key containing the catalog_id and product_id 
+enables us to easily update the status across one, all, or specific catalogs.
+
+CatalogProductSetStatusAction does not set the status for multiple 
+catalog_products. It will only change the status of a single 
+catalog_product. The nuance being that this status change can be reflected in 
+as many, or as few catalogs an entity desires.
+
+This flexibility is to enable support for use cases like:
+
+- I am discontinuing a catalog_product for entity A, B, C, but not for retailer D.
+- I want to discontinue catalog_products for all entity A, B, C, and D.
+- I am marking a catalog_product as active for entity A and B, but not C or D just yet.
+- I am making an active catalog_product inactive for entity A and B, but not C or D.
+
+```
+message CatalogProductSetStatusAction {
```

Comment:
So this transaction operates on a catalog product schema. Couldn't there be a case where one of the specified catalog product schemas don't have  the `status` or `status_change_reason` PropertyValues? What would happen in that case?

Also, couldn't you just accomplish this by using the `CatalogProductUpdateAction` transaction, which doesn't make any assumptions about the `PropertyValues`?

Reply:
The discussion that led to including `status` as a required field of a catalog_product schema was to show that Grid has some opinion of what should be included in a catalog_product. Since catalog_products are intended to be used for business partner-specific relations, we landed on a `status` field (indicating a catalog_products availability) that is something to be shared among all catalog_products.


From an implementation standpoint, this could certainly be implemented as a `CatalogProductUpdateAction`. As I think more about it, I don't see too strong a reason to have this as its own transaction. I'm trying to recall the reasoning of implementing it as a separate transaction, maybe be to enforce the significance of the `status` (availability) field. 

Do you recall the reasoning @davececchi @genasmorgan? You'd be able to explicitly see this transaction on a protocol level, not what the value prop was of that. If there is no strong opposition I'm open to scraping the `CatalogProductSetStatusAction` and just batching it in with `CatalogProductUpdateAction`.

Reply:
If Grid is opinionated about what is in the Catalog Product data model, I would recommend making that its own protobuf and including the status as a field. Otherwise, I don't think it would be ideal to have transactions that explicitly reference specific data that isn't included in the message.

Reply:
Good point. I think we want to be opinionated, but not forceful. There will be a schema example with status (among other fields) coming out of this RFC. That gives developers a base schema to start with but doesn't force them to use any specific fields. That being said, i'll remove the`CatalogProductSetStatusAction` and that will get batched in with `CatalogProductUpdateAction`

Reply:
Actually, as I was going to remove the `CatalogProductSetStatusAction` it became clear to me why we needed the `CatalogProductSetStatusAction`. 

In order to set the status of a catalog_product via the CatalogProductUpdateAction you'd need to overwrite all the properties for all the additional fields of that specific catalog, also to do this across more than one catalog, you'd need a separate transaction for each catalog_product. So If I were to discontinue a catalog_product, across 100 catalogs that would require 100 CatalogProductUpdateAction's. The purpose of `CatalogProductSetStatusAction`is to conveniently batch those related updated into one transaction.

As for the enforcement of the `status` field, the smart contract could throw an error if the catalog_product does not have a `status` field in its additional fields. I want to avoid creating a `catalog_product` message just for enforcing one field. The reusing of Grid Product would get us the functionality we need, without the overhead of managing an additional message type that is only slightly different from the base.

Reply:
Ok. I added a PR to your branch expressing that alternative.